core/src/main/java/com/linecorp/armeria/server/RoutingTrie.java

```diff
@@ -564,4 +502,167 @@ static void validatePath(@Nullable String path) {
                           Integer.toHexString(KEY_CATCH_ALL), path);
         }
     }
+
+    /**
+     * A strategy to find values in {@link RoutingTrie}.
+     */
+    interface FindStrategy {
+
+        static FindStrategy findFirst() {
+            return FindFirstStrategy.INSTANCE;
+        }
+
+        static FindStrategy findAll() {
+            return FindAllStrategy.INSTANCE;
+        }
+
+        <V> List<V> find(Node<V> node, String path, int begin, boolean exact);
```

Comment:
Just a thought - I'm curious if we can parameterize `FindStrategy` so that it returns `V` or `List<V>`:

```java
private interface FindStrategy<T> {
    static FindStrategy<T> findFirst() { ... }
    static FindStrategy<List<T>> findAll() { ... }
}
```

... or maybe we could just give up introducing abstraction here and have two distinct implementations? i.e. `find()` and `findAll()` share no logic.

Reply:
Both of two options look good to me. Now, I'm thinking how to revamp `FindStrategy` based on your advice https://github.com/line/armeria/pull/2040#discussion_r324577154. 🧐

Reply:
> we could just give up introducing abstraction here and have two distinct implementations? 

I wonder IIUC 🤔
* Remove `FindStrategy`, revert this to the master branch code.
* Implement new method `findAllNode(...)` which copies logic code from `findNode(...)` and share no logic.
https://github.com/line/armeria/blob/d89c801548d13de6af64cd3655a2e083fbfbed5c/core/src/main/java/com/linecorp/armeria/server/RoutingTrie.java#L107-L109

```java
final class RoutingTrie<V> {

    ...

    @Nullable
    List<V> find(String path) {
        final Node<V> node = findNode(path, false);
        return node == null ? ImmutableList.of() : node.values();
    }

    List<V> findAll(String path) {
        final List<Node<V>> nodes = findAllNode(path, false);
        return nodes.stream().flatMap(...).toList(...);
    }
  
    @Nullable
    private Node<V> findNode(Node<V> node, String path, int begin, boolean exact) {
        ...
    }
  
    @Nullable
    private List<Node<V>> findNode(Node<V> node, String path, int begin, boolean exact) {
        ...
    }

}
```

Reply:
+1 for `two distinct implementations`.
I think we can just use static methods which share `checkNode()`.

Reply:
`FindStrategy` is gone. ⚰️